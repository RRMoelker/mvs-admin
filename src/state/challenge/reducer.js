import { createAction } from 'redux-actions';

// import { reduxSwarmLogId } from '../actions.js';

import { v4 } from 'uuid';

import { mergeOverlapping } from './merger.js';
import {
    calculateActive,
    calculateRecent,
    calculateRemaining,
    groupChallenges,
    getRemaining
} from './compute.js';

const CHALLENGE_ADD = 'CHALLENGE_ADD';
// export const addChallenge = createAction(CHALLENGE_ADD);
export const addChallenge = payload => ({
    type: CHALLENGE_ADD,
    payload,
    // reduxSwarmLogId
});

const CHALLENGE_REMOVE = 'CHALLENGE_REMOVE';
// export const removeChallenge = createAction(CHALLENGE_REMOVE);
export const removeChallenge = payload => ({
    type: CHALLENGE_REMOVE,
    payload,
    // reduxSwarmLogId
});

const initialState = {
    list: [],
}

export const selectRemaining = (list, now) => {
    const grouped = groupChallenges(list);

    const remaining = {};
    for( const [ key, groupList ] of Object.entries(grouped)) {
        const merged = mergeOverlapping(groupList);
        const active = calculateActive(merged, now);
        if ( active.length ) {
            remaining[key] = calculateRemaining(active, now);
        }
    }
    return remaining;
}

export const selectRecent = (list, now, threshold) => {
    const recentList = calculateRecent(list, now, threshold);
    const grouped = groupChallenges(recentList);
    const result = {};
    for( const [ key, value ] of Object.entries(grouped) ) {
        result[key] = value.reduce(
            (acc, val) => acc + val.duration,
            0
        );
    };
    return result;
}

export default (state = initialState, action) => {
    switch (action.type) {
        case CHALLENGE_ADD:
            const { name, duration } = action.payload;
            const { time } = action.meta;

            const newChallenge = {
                uuid: v4(),
                name,
                duration,
                time,
            };
            const list = [ ...state.list, newChallenge];
            list.sort((a,b) => {
                return a.time > b.time;
            });

            return {
                ...state,
                list,
            }
        case CHALLENGE_REMOVE:
            const filteredList = state.list.filter( item => item.uuid !== action.payload.uuid);
            return {
                ...state,
                list: filteredList
            };
        default:
            return state;
    }
};
