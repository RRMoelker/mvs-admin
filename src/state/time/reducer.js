import { createAction } from 'redux-actions';

export const TIME_SET = 'TIME_SET';
export const setTime = createAction(TIME_SET);
const TIME_RESET = 'TIME_RESET';
export const resetTime = createAction(TIME_RESET);
const TIME_PAUSE = 'TIME_PAUSE';
export const pauseTime = createAction(TIME_PAUSE);

const initialState = {
    time: 0,
    running: true
}

export default (state = initialState, action) => {
    switch (action.type) {
        case TIME_RESET:
            return {
                ...state,
                time: 0,
            }
        case TIME_SET:
            return {
                ...state,
                time: action.payload.time,
            }
        case TIME_PAUSE:
            return {
                ...state,
                running: !state.running
            }
        default:
            return state;
    }
}
