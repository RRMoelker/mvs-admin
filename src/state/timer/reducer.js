import { createAction } from 'redux-actions';

export const TIMER_SET = 'TIMER_SET';
export const setTimer = createAction(TIMER_SET);
const TIMER_RESET = 'TIMER_RESET';
export const resetTimer = createAction(TIMER_RESET);
const TIMER_PAUSE = 'TIMER_PAUSE';
export const pauseTimer = createAction(TIMER_PAUSE);

const initialState = {
    time: 0,
    running: true
}

export default (state = initialState, action) => {
    switch (action.type) {
        case TIMER_RESET:
            return {
                ...state,
                time: 0,
            }
        case TIMER_SET:
            return {
                ...state,
                time: action.payload.time,
            }
        case TIMER_PAUSE:
            return {
                ...state,
                running: !state.running
            }
        default:
            return state;
    }
}
