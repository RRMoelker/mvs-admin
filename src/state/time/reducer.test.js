import reducer, {
    setTime,
    resetTime,
    pauseTime
} from './reducer.js';


it('should set the time', () =>{
    const result = reducer({
        time: 5
    }, setTime(2));

    expect(result).toMatchSnapshot();
});

it('should reset the time', () =>{
    const result = reducer({
        time: 5
    }, resetTime());

    expect(result).toMatchSnapshot();
});

it('should pause the timer', () => {
   const result = reducer({
       running: false
   }, pauseTime());

   expect(result).toMatchSnapshot();
})

it('should resume the timer', () => {
   const result = reducer({
       running: false
   }, pauseTime());

   expect(result).toMatchSnapshot();
})
