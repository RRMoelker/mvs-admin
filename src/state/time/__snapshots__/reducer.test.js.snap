// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should pause the timer 1`] = `
Object {
  "running": true,
}
`;

exports[`should reset the time 1`] = `
Object {
  "time": 0,
}
`;

exports[`should resume the timer 1`] = `
Object {
  "running": true,
}
`;

exports[`should set the time 1`] = `
Object {
  "time": 2,
}
`;
